### ISSUE TITLE
Test fails comparing two identical list of tuples

### ISSUE THREAD

**Issue by dracan** [xunit/xunit#1995]
This test fails, but I'm not sure why...

```cs
[Fact]
public void Test1() =>
    Assert.Equal(("Some string", new List<(string key, string value)> { ("Name", "Bob") }),
        ("Some string", new List<(string key, string value)> { ("Name", "Bob") }));
```

If I use a tuple without a list, it passes...

```cs
[Fact]
public void Test2() =>
    Assert.Equal(("Some string", (name: "Key", value: "Bob")),
        ("Some string", (name: "Key", value: "Bob")));
```

**Comment by OrbintSoft** [comment]
because the two list are two different objects, even if they contain the same elements.
Tuple implements IComparable, and it's possible to compare them, if they are a tuple of primitive types like strings in this case,
Anyway this is not a xunit bug, but it the normal behavior of .NET, xunit only calls Object.Equals(a,b)
Ask support on stackoverflow if you need.

**Comment by Kerl1310** [comment]
@dracan This looks like it could be due to the hashcode of the lists being different... Can you move the List initializations out of the Assert.Equal and then check the hashcodes to confirm?

**Comment by dracan** [comment]
@Kerl1310 - sorry for the ridiculous (2-year!) delay in my reply. Just noticed I hadn't replied back to your question. The hashcodes are different - but saying that, they're also different for identical lists that don't use tuples too - so I don't think this is related...

![image](https://user-images.githubusercontent.com/567988/128622655-4072041b-f1e2-41d3-8e7b-109affb3330d.png)

**Comment by bradwilson** [comment]
The problem is that tuples implement `IEquatable<>` and when we see classes that implement that, we assume that to be "the truth". Their implementation for equality is straightforward:

```csharp
public bool Equals((T1, T2) other)
{
    if (EqualityComparer<T1>.Default.Equals(Item1, other.Item1))
    {
        return EqualityComparer<T2>.Default.Equals(Item2, other.Item2);
    }
    return false;
}
```

Suffice to say, `EqualityComparer<>.Default` is not what you'd want to use in a unit test, because in the case of reference types, you're basically getting reference equality. Because the two lists are _different instances_, `EqualityComparer<>` says "they're not equal".

In short, Tuple claims to understand equality, and it's not what you want.

There's not much we can do here short of adding in custom code to detect the use of tuples and telling it to ignore the implementation of `IEquatable<>`. I'm reluctant to do that, but I'll leave the issue open for discussion to see how others feel about it.

**Comment by bradwilson** [comment]
Without any further feedback on this, I'm going to close this as won't fix at this point in time.
